pkg/maintenancemode/config_test.go

```diff
@@ -0,0 +1,135 @@
+	corev1clientmocks "github.com/SAP/stewardci-core/pkg/k8s/mocks/client-go/corev1"
+	gomock "github.com/golang/mock/gomock"
+	"github.com/pkg/errors"
+	"gotest.tools/assert"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/system"
+	_ "knative.dev/pkg/system/testing"
+)
+
+func Test_IsMaintenanceMode_getError_(t *testing.T) {
+	t.Parallel()
+	// SETUP
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	cf := newErrorFactory(mockCtrl, api.MaintenanceModeConfigMapName, errors.New("some error"))
+
+	// EXERCISE
+	_, resultErr := IsMaintenanceMode(cf)
+
+	// VERIFY
+	assert.Error(t, resultErr, `invalid configuration: ConfigMap "steward-maintenance-mode" in namespace "knative-testing": some error`)
+}
+
+func Test_IsMaintenanceMode_get_NotFoundError(t *testing.T) {
+	t.Parallel()
+	// SETUP
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	cf := newErrorFactory(mockCtrl, api.MaintenanceModeConfigMapName, k8serrors.NewNotFound(api.Resource("pipelineruns"), ""))
+	// EXERCISE
+	result, resultErr := IsMaintenanceMode(cf)
+
+	// VERIFY
+	assert.Assert(t, result == false)
+	assert.NilError(t, resultErr)
+}
+
+func Test_IsMaintenanceMode_configMapHasDeletionTimestamp(t *testing.T) {
+	t.Parallel()
+	// SETUP
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	cm := newControllerConfigMap(map[string]string{
+		"maintenanceMode": "true",
+	})
+	cm.ObjectMeta.DeletionTimestamp = &metav1.Time{Time: time.Now()}
+	cf := fake.NewClientFactory(cm)
+	// EXERCISE
+	result, resultErr := IsMaintenanceMode(cf)
+
+	// VERIFY
+	assert.Assert(t, result == false)
+	assert.NilError(t, resultErr)
+}
+
+func Test_loadControllerConfig(t *testing.T) {
+	t.Parallel()
+	for _, tc := range []struct {
+		name       string
+		configData map[string]string
+		expected   bool
+	}{
+		{
+			"MaintenanceModeEnabled",
+			map[string]string{
+				"maintenanceMode": "true",
+			},
+			true,
+		},
+		{
+			"MaintenanceModeDisabled",
+			map[string]string{
+				"maintenanceMode": "false",
+			},
+			false,
+		},
+		{
+			"MaintenanceModeMissing",
+			map[string]string{},
+			false,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			tc := tc // capture current value before going parallel
+			t.Parallel()
+			// SETUP
+			cf := fake.NewClientFactory(
+				newControllerConfigMap(tc.configData),
+			)
+
+			// EXERCISE
+			result, resultErr := IsMaintenanceMode(cf)
+
+			// VERIFY
+			assert.NilError(t, resultErr)
+			assert.Equal(t, tc.expected, result)
+		})
+	}
+}
+
+func newControllerConfigMap(data map[string]string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      api.MaintenanceModeConfigMapName,
+			Namespace: system.Namespace(),
+		},
+		Data: data,
+	}
+}
+
+func newErrorFactory(mockCtrl *gomock.Controller, configMapName string, expectedError error) *mocks.MockClientFactory {
```

Comment:
The function name is too generic and misleading. According to what it does, I would name it `newClientFactoryWithConfigMapIfceExpectingGetAndReturnsError`. This is an indicator that the function combines too many things and needs a refactoring.

You could have one function creating a _basic_ client factory mock only, and add the specific config map mocks directly in your test function or via a separate function (if needed more than once). That way the test function would be better understandable.

Reply:
See ff86204b502c87ca2c479dd923ac157d7ed9c839

Reply:
Yes, that's better.